feat: Introducing Tracing with OpenTelemetry API

jimit-j-shah opened the pull request:
This change introduces Client-Side Tracing using OpenTelemetry API in Datastore Java SDK.

User Guide & Documentation: 

- https://cloud.google.com/datastore/docs/client-side-traces
- https://cloud.google.com/datastore/docs/trace-span-references

Fixes googleapis/java-datastore#1428 ☕️

jimit-j-shah committed:
feat: Adding TraceUtil interface and its implementation to enable Tracing controls via DatastoreOptions (#1431)

* Adding EnabledTraceUtil, DisabledTraceUtil and TraceUtilTest
* Annotating DatastoreOpenTelemetryOptions to be transient as they're not serializable
* Adding google-auth-library-credentials dependency due to https://github.com/googleapis/java-datastore/actions/runs/8944472794/job/24571458116?pr=1431

jimit-j-shah committed:
feat: Adding Lookup RPC OpenTelemetry Tracing (#1437)

* feat: Adding Lookup RPC OpenTelemetry Tracing
- Removed OpenCensus Tracing
- Added E2E tests with Global and Local OTel SDK
- Moved OTel SDK setup to RemoteDatastoreHelper
- Fixed pom to depend on BOM for all shared dependencies

jimit-j-shah committed:
feat: RunQuery trace instrumentation

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
Refactor: s/RUNQUERY/RUN_QUERY

jimit-j-shah committed:
feat: RunAggregationQuery Trace Instrumentation

jimit-j-shah committed:
Build: retiring test assertions for OpenCensus spans - will be replacing this in hermetic integration tests for OpenTelemetry using in-memory span exports (in addition to ITE2ETraceTest.java).

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
feat: Adding Commit RPC Trace Instrumentation (#1440)

- Added end-to-end test for Datastore operationsput, add, update and delete.
- Updated E2E Test to use the namespace correctly for efficient clean-up of test data

jimit-j-shah committed:
Fixing @Test annotation missed after merge

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
feat: Add Transaction tracing
test: transactionalLookupTest

jimit-j-shah committed:
test: Transaction test for RunInTransaction - need to fix trace instrumentation for RunIn..

jimit-j-shah committed:
feat: RunQuery trace instrumentation (#1441)

* feat: RunQuery trace instrumentation

jimit-j-shah committed:
feat: RunAggregationQuery instrumentation (#1447)

* feat: RunQuery trace instrumentation

* Formatting

* Formatting

* Refactor: s/RUNQUERY/RUN_QUERY

* feat: RunAggregationQuery Trace Instrumentation

* Build: retiring test assertions for OpenCensus spans - will be replacing this in hermetic integration tests for OpenTelemetry using in-memory span exports (in addition to ITE2ETraceTest.java).

* Formatting

* Fixing @Test annotation missed after merge

* Formatting

jimit-j-shah committed:
Adding transaction span names

jimit-j-shah committed:
TransactionLookupTest

jimit-j-shah committed:
feat: support for transactional operations
- tested using newTransaction() and runInTransaction()

jimit-j-shah committed:
Revert "feat: support for transactional operations"

This reverts commit 10341c0b97cbc2025f9f928ce8cb09d5c036a5b3.

jimit-j-shah committed:
feat: support for transactional operations (#1468)

* feat: support for transactional operations
- tested using newTransaction() and runInTransaction()

jimit-j-shah committed:
feat: Allocateid tracing (#1488)

* feat: Adding tracing for AllocateIds RPC

* formatting

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

jimit-j-shah committed:
feat: Add tracing for ReserveIds operation (#1490)

- added end-to-end test

jimit-j-shah committed:
fix: Fixed Span nesting for `ReadWriteTransactionCallable` by using parent SpanContext instead of just parent Context (#1495)

* fix: Fixed the TraceUtil.startSpan method to use `SpanContext` for linking with the parent instead of `Context`.
- This fixes the hierarchy of Spans appearing in a transaction under a Run method.
- Tested using existing transaction test

* Fixed commit reordering and typos

* fix: lint errors

* fix: Refactored the ReadWriteTransactioncallable.call method to use startSpan idiomatically
- TraceUtil.startSpan needs more debugging
- return DefaultTracerProvider instance (no-op) when initializing DisabledTraceUtil - this fixes the unit tests in DatastoreTest.testRunInTransactionWithReadWriteOption

* feat: Added tracing for Transaction.RunQuery (#1499)

* feat: Added span for Transactional RunQuery
- tested

* fix: lint

* fix: patch apply issues

* fix: refactor using boolean flag

* fix: s/startSpan/startSpanWithParentContext

jimit-j-shah committed:
test: Additional Transaction Testing and cleanup OpenCensus usage (#1505)

* test: newTransactionReadWriteTraceTest

* fix: test literal

* feat: Added tests for transaction cases

* fix: Cleanup OpenCensus dead code

* fix: updating version from 2.20.1 -> 2.21.0

* fix: reverting version from 2.21.0 -> 2.20.1

* fix: Adding an exception to the clirr-maven-plugin for an internal API parameter change from com.google.cloud.datastore.TraceUtil -> com.google.cloud.datastore.telemetry.TraceUtil

* fix: Fixing the differenceType in clirr exception

* fix: add an exception for removal of an internal class (com.google.cloud.datastore.TraceUtil)

* fix: fixing incomplete difference details for type 7005

* fix: Fixing `to` of the difference to be the entire signature

* fix: typo

jimit-j-shah committed:
test: Adding ITTracingTest to verify events and span attributes (whic… (#1514)

* test: Adding ITTracingTest to verify events and span attributes (which are not verified in ITE2ETracingTest) due to TraceClient API limitations.
- This test uses InMemorySpanExporter to read the generated Otel span data by the test process to verify generated span data as it were before exporting to a backend. None of the span data is exported to a durable backend.
- This test is still an E2E test as it requires a project to send RPCs to.

* fix: fixing compilation error due to missing pom dependency.

* test: Test for AllocateId and ReserveId rpcs

* test: Commit/Put/Update/Delete tests

* test: Added fixes and test for RunQuery event

jimit-j-shah committed:
test: Additional Transaction tests and AggregationQuery test (#1518)

* test: ReadWrite Transaction test

* test: Added test for Transactional RunQuery and Transaction Rollback

* test: runInTransaction API tracing test
- Fixed setting of common span attributes to spans in runInTransaction
- Removed some gRPC related channel attributes that are not present in this Datastore version, yet.

jimit-j-shah committed:
fix: Undelete gRPC upgrade docs

jimit-j-shah committed:
fix: Undo merge mistakes

release-please[bot] committed:
chore(main): release 2.21.0 (#1517)

:robot: I have created a release *beep* *boop*
---


## [2.21.0](https://togithub.com/googleapis/java-datastore/compare/v2.20.2...v2.21.0) (2024-07-31)


### Features

* Enable hermetic library generation ([#1462](https://togithub.com/googleapis/java-datastore/issues/1462)) ([d142d9c](https://togithub.com/googleapis/java-datastore/commit/d142d9c95d91c8cadaf696efc12d6136814938ff))

---
This PR was generated with [Release Please](https://togithub.com/googleapis/release-please). See [documentation](https://togithub.com/googleapis/release-please#release-please).

jimit-j-shah committed:
fix: Updating span event strings (#1539)

* fix: Fixing user-facing span names in line with go/firestore-client-trace-catalog

* fix: updating bom dependency version to fix https://github.com/googleapis/java-datastore/actions/runs/10256441634/job/28375496112?pr=1539

renovate-bot committed:
deps: update dependency com.google.cloud:sdk-platform-java-config to v3.33.0 (#1531)

release-please[bot] committed:
chore(main): release 2.21.1-SNAPSHOT (#1538)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

renovate-bot committed:
chore(deps): update dependency com.google.cloud:libraries-bom to v26.43.0 (#1515)

cindy-peng commented:
Hi @jimit-j-shah, is this pull request combining all previous changes made on the feature branch`[tracing-main-3](https://github.com/googleapis/java-datastore/tree/tracing-main-3)`?

jimit-j-shah commented:
> Hi @jimit-j-shah, is this pull request combining all previous changes made on the feature branch [tracing-main-3](https://github.com/googleapis/java-datastore/tree/tracing-main-3)?

Yes. Just rebased on `main`'s new changes into a new branch.

jimit-j-shah commented:
Will wait to merge until we have launch approval.

release-please[bot] committed:
chore(main): release 2.21.1 (#1540)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

release-please[bot] committed:
chore(main): release 2.21.2-SNAPSHOT (#1541)

:robot: I have created a release *beep* *boop*
---


### Updating meta-information for bleeding-edge SNAPSHOT release.

---
This PR was generated with [Release Please](https://togithub.com/googleapis/release-please). See [documentation](https://togithub.com/googleapis/release-please#release-please).

diegomarquezp committed:
chore: secure hermetic_library_generation workflow (#1552)

* chore: secure hermetic_library_generation workflow

Thanks to @diogoteles08 for the inspection on our repos. This PR inlines environment variables to avoid overriding script injections.

* fix wording

renovate-bot committed:
deps: update dependency com.google.cloud:sdk-platform-java-config to v3.34.0 (#1547)

release-please[bot] committed:
chore(main): release 2.21.2 (#1553)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

mpeddada1 committed:
chore: remove datastore native image sample in favor of sample hosted in google-cloud-java (#1519)

jimit-j-shah committed:
Fix: typo in test causing integration test failure (#1556)

https://btx.cloud.google.com/invocations/c11a2e8b-4494-4ddc-a77e-cf2bcbcf5254/targets/cloud-devrel%2Fclient-libraries%2Fjava%2Fjava-datastore%2Fpresubmit%2Fintegration;config=default/log

cindy-peng commented on .github/workflows/hermetic_library_generation.yaml:
```diff
@@ -17,10 +17,14 @@ name: Hermetic library generation upon generation config change through pull req
 jobs:
   library_generation:
     # skip pull requests coming from a forked repository
-    if: github.event.pull_request.head.repo.full_name == github.repository
```
Could we separate these unrelated commits?

jimit-j-shah replied:
I never made these commits. Seem to be rebase/merge artifacts?

cindy-peng commented on google-cloud-datastore/pom.xml:
```diff
@@ -160,6 +194,56 @@
+    <!-- END OpenTelemetry -->
+    <!-- Cloud Ops -->
+    <dependency>
+      <groupId>com.google.cloud.opentelemetry</groupId>
```
Nit: Move this before `<!-- End OpenTelemetry -->`

jimit-j-shah replied:
This is a cloud dependency, not an opentelemetry dependency.

cindy-peng commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:
```diff
@@ -125,20 +142,57 @@ void setPrevTransactionId(ByteString transactionId) {
     public T call() throws DatastoreException {
-      transaction = datastore.newTransaction(options);
-      try {
+      // TODO Instead of using OTel Spans directly, TraceUtil.Span should be used here. However,
```
Do we have a github issue tracking this?

jimit-j-shah replied:
#1491

ehsannas commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:
```diff
@@ -602,20 +692,37 @@ private com.google.datastore.v1.CommitResponse commitMutation(
+              .put(
+                  "transaction_id",
+                  isTransactional ? requestPb.getTransaction().toStringUtf8() : "")
+              .build());
```
nit: It'd be better to have all attribute keys defined as constants somewhere (similar to [this](https://github.com/googleapis/java-firestore/blob/main/google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/TraceUtil.java#L58-L65))

jimit-j-shah replied:
Done.

ehsannas reviewed (commented):
The public APIs look good to me. I also briefly reviewed certain parts of the PR (skipped the tests). I've pointed out some nits below. I defer the approval to the team that maintains this repo.

jimit-j-shah commented:
> Hi @jimit-j-shah, is this pull request combining all previous changes made on the feature branch [tracing-main-3](https://github.com/googleapis/java-datastore/tree/tracing-main-3)?

Yes. Now I've done a fresh merge from main in https://github.com/googleapis/java-datastore/pull/1576. Let's continue the conversation there.

jimit-j-shah commented:
> Will wait to merge until we have launch approval.

We now have launch approval.

jimit-j-shah commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java:
```diff
@@ -40,6 +39,7 @@
 import com.google.datastore.v1.RunAggregationQueryResponse;
 import com.google.datastore.v1.RunQueryRequest;
 import com.google.datastore.v1.RunQueryResponse;
+import io.opencensus.trace.Tracing;
```
Replaced `TraceUtil` usage https://github.com/googleapis/java-datastore/commit/45da20b4a502c2bbe5b03d52f843ce5e39174c76

jimit-j-shah commented on google-cloud-datastore/pom.xml:
```diff
@@ -111,14 +128,31 @@
+    <!-- OpenTelemetry -->
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-sdk</artifactId>
```
Typo - it should have been under `Test Dependencies` - good catch!

jimit-j-shah commented:
This PR is superseded by https://github.com/googleapis/java-datastore/pull/1576 - will implement all the comments there.

jimit-j-shah commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:
```diff
@@ -125,20 +142,57 @@ void setPrevTransactionId(ByteString transactionId) {
+      // probably because of some thread-local caching that is getting lost. This needs more
+      // debugging. The code below works and is idiomatic but could be prettier and more consistent
+      // with the use of TraceUtil-provided framework.
+      io.opentelemetry.api.trace.Span span =
```
- Looks like ReadWriteTransactionCallable is scheduled to run on an Executor in the `RetryHelper` library provided by gax layer. So that complicates the Context propagation.
- Also tried "manually" propagating the parent SpanContext to the "callable" but that doesn't work either as we don't have control over the future lifetimes (which run in the RetryHelper) so the parent span may `end()` before the child spans are executed(?).
- This needs some significant overhauling or gax support for context propagation - both are non trivial and wouldn't want to block the launch on this.

ehsannas replied:
Uh... not sure why using the TraceUtil directly doesn't work. Maybe it's because there's only 1 TraceUtil but several threads creating spans? If that's the case, can you try and see if this works:

Add an API `TraceUtil.isTracingEnabled()`, and then change:

```java
private TraceUtil.Span startSpanWithParentContext(
        String spanName,
        TraceUtil.SpanContext parentSpanContext) {
  TraceUtil otelTraceUtil = datastore.getOptions().getTraceUtil();
  if (!otelTraceUtil.isTracingEnabled()) {
    return new TraceUtil.Span();
  }

  SpanBuilder spanBuilder =
      otelTraceUtil
          .getTracer()
          .spanBuilder(com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN)
          .setSpanKind(SpanKind.PRODUCER)
          .setParent(
              Context.current()
                  .with(
                      io.opentelemetry.api.trace.Span.wrap(
                          parentSpanContext.getSpanContext())));
  return TraceUtil.Span(otelTraceUtil.addSettingsAttributesToCurrentSpan(spanBuilder).startSpan());
}
```

```java
TraceUtil.Span span =  startSpanWithParentContext(
              com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN,
              parentSpanContext);

try (TraceUtil.Scope ignored = span.makeCurrent()) {
...
```

This way the opentelemetry code will remain off the execution path if tracing is disabled, and OTEL types are also not used on the main code path.
